### ISSUE TITLE
Connecting to matrix with android element over mobile networks takes 40-70 seconds

### ISSUE THREAD

**Issue by KeenMaron** [vector-im/element-android#4618]
### Steps to reproduce

Hi!

I was successfully able to set up Matrix on my RaspberryPi 4 months ago.

I use the Element Client on Windows and on my mobilephone (Android).

The connection in the same network is established nearly (2-3 seconds) immediately.
The connection is also established very quickly when I log in with a Windows PC from another network.

Two friends of mine have also connected with their mobilephone without any signifcant delay problems (iphone and a android phone). They used their mobile network or own cable network from home to establish the first connection.

I have connected my phone initially from the WiFi network (same Network like Matrix Server).

As soon as I leave the WiFi and try to connect via mobile network, Element needs 40-70 seconds to establish the connection. Then it works as long as the app is open.

I've had this problem before and at some point the connection-delay just disappeared. Now I have changed the port 8448 to another port, and the problem occurs again.

My girlfriend's phone also has the problem. She also initialy logged from the same network as Matrix, like me. When she switches to the mobile network the establishment for an connection takes about 30 seconds (definitely less than my phone).

What i actually figured out:

- Connection via HotSpot from phone (mobile network) to Windows PC
-- Element Windows Client is logging in without any big delay
- Connection HotSpot from phone (mobile network) to Android Tablet using Android Element
-- Big delay appears again
- I tried the android App FluffyChat. Initial connection via mobile network had nearly no delay. The delay appears only with the Element Android app
- Reinstalling Element (Beta and non Beta), same behaviour :(

Do you have any advice what might be the problem?

#### What did you expect?
Connecting via mobile network should take few seconds - like in FluffyChat (also an Android Matrix Client App)

#### What happened instead?
Establishing connection via mobile network takes 40-70 seconds.

### Your phone model

s9+ and Xiaomi RedMi Note 10

### Operating system version

Android 10

### Application version and app store

Element Version 1.3.7 from Google App Store

### Homeserver

selfhosted matrix server (raspberryPi) with dyn dns

### Will you send logs?

No

**Comment by ouchadam** [comment]
@KeenMaron thanks for raising, this sounds very strange... other than prefetching extra information when receiving a push notification there's no other logic which is network dependant  (at least that I'm aware of!) 

something to bear in mind is that synapse does sync caching which could cause some scenarios to feel faster than they might actually be

----

I have a few questions to help debug this! 

- are the delay issues intermittent? 

- when you say `connection-delay`, what's happening on the screen? when the progress bar at the top of screen is displaying we're executing an incremental sync request

![Screenshot_20211202_164020](https://user-images.githubusercontent.com/1848238/144465179-8fd1f74c-3ffc-42dd-b5b8-e071b5e4f028.png)

- if possible are you able to have FluffyChat on one device and Element on another, logged in to the same account, receiving messages to see if there's a large difference in timings when both are under the same conditions

- does the app show the same slowness when using the mobile network and matrix.org accounts?

- and finally... how is the RPi's CPU usage whilst the big delays are happening? you could have a user on the homeserver who is in large rooms causing slow downs

**Comment by KeenMaron** [comment]
Hi @ouchadam ! Thank you for your fast reply. I will test all the cases and answer as soon as possible :).

**Comment by KeenMaron** [comment]
**Are the delay issues intermittent?**
-	Actually no. There is always a delay if I connect through my mobile network. No delay with Wifi connection (same network as matrix server).

**when you say connection-delay, what's happening on the screen? when the progress bar at the top of screen is displaying we're executing an incremental sync request**
-	Yes. The progress bar on the top is working ~40-70 seconds. If I try to send messages in that period, the receiver does not get any messages. When the progress bar disappears, then all the messages flush immediately out. There is no delay after the “sync request” between sending and receiving messages.

**if possible are you able to have FluffyChat on one device and Element on another, logged in to the same account, receiving messages to see if there's a large difference in timings when both are under the same conditions**
-	Already tested. There is no notable difference in timings receiving / sending messages. Only the initial connection takes much more time at Element (Element: 40-70 seconds; Fluffy: 2-3 seconds) after opening the apps. 

**does the app show the same slowness when using the mobile network and matrix.org accounts?**
-	Good point. I just started to register an account at matrix.org, but stopped because I noticed one thing (everything tested with mobile network):
-	If I choose matrix.org server, a loading “donut” occurs for a few seconds. After that i can login / register.
-	If I choose custom server and type the address of my server (https:\\myservername.pi.net:port) the loading “donut” takes 65 seconds to the login / register screen. Really weird. I monitored the cpu and memory usage of my Pi the whole time. No spikes, no cpu load over 7%. Memoryusage 15%.
-	After that I finished my registration at matrix.org. Login was no problem. I saw the progress bar at the top for 1-2 seconds.
-	I logged off and logged into my server with my accout. Loading took 1 minute, after that I typed my credentials and the sync progress (data, … , verification) took another minute

**and finally... how is the RPi's CPU usage whilst the big delays are happening? you could have a user on the homeserver who is in large rooms causing slow downs**
-	Good point. But: I and my girlfriend never joined any rooms. No spikes in CPU and Memory usage. I watched it in realtime with htop.
-	After logging in and out in element and fluffychat + several (cross)verifications processes I noticed that fluffyChat in one of two cases took 5-10 seconds longer to establish a connection over mobile network. Maybe a kind of overhead with verification processes?

My guess is, that my matrix server does not respond correctly to the “incremental sync requests”? 
Maybe FluffyChat does not requests some functionalities like Element, and that is the reason why it establishes the connection so fast?

Sidenote:
I thought the reason could be the “online-status” flag in the matrix config. FluffyChat has no user online-status. I disabled it and rebooted the server = still same behaviour :(

**Comment by ouchadam** [comment]
and to double double check, whilst on WiFi these delays aren't as bad?

something that can help debug from our side is temporarily enabling verbose logs (Settings -> Advanced Settings -> Enable developer mode -> Enable verbose logs) and then sending the logs the 3 dots menu in the home screen `report a bug`

these verbose logs will include a timing breakdown of the (incremental and initial syncs) https://github.com/vector-im/element-android/blob/develop/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/SyncTask.kt#L138

if you could go through the login flow and wait for the loading to finish then report for both Network and WiFi then we should be able to determine where the delay is coming from.

If the delay is coming from the network request then it suggests either Element android is requesting the data differently to the other clients or there's a potential network configuration issue.

**Comment by KeenMaron** [comment]
**and to double double check, whilst on WiFi these delays aren't as bad?**
Correct, nearly no delay (2-3 seconds) while logging in through Wifi. If i use VPN to my home network thorugh mobile network, the establishment of the connection takes 2-3 seconds too. I could try another wifi network -> i will give you feedback as soon i have access to another network (maybe today in the evening) 

**verbose logs**
Ok thanks, i will try that later or tomorrow. Do I have to remove some sensible data from the logs? Probably the server adress, i think?

**Comment by ouchadam** [comment]
thanks for the help, no rush :smile: 

the logs sent by the `report bug` screen are uploaded to our private server, they don't include any message contents but they do include user ids and homeserver addresses. If you would prefer to redact this information unfortunately it will mean plugging in your android device and manually pulling the logs off via adb logcat https://thedarksource.com/adb-logcat-command-to-capture-debug-logs-from-android-device/ and then editing the logs and uploading to pastebin or including in a comment here (would recommend using the `report bug` instead!)

>  If i use VPN to my home network through mobile network, the connection takes 2-3 seconds too

this has me leaning towards some sort of network issue, trying another WiFi network would help narrow down some of the potential cases. Perhaps there's a firewall causing external traffic to be delayed (but that wouldn't explain the other clients working...)

**Comment by KeenMaron** [comment]
**this has me leaning towards some sort of network issue, trying another WiFi network would help narrow down some of the potential cases. Perhaps there's a firewall causing external traffic to be delayed (but that wouldn't explain the other clients working...)**

I was few hours ago in a restaurant with wifi. I disabled the mobile network first, connected with the wifi and started element. The connection was immediately initialized. No delay. After that I switched to mobile network and again 1-2 minutes delay.
I tested my mobile network with this site https://test-ipv6.com/ I got an ipv4 and 6 address. I the wifi of the restaurant i only had ipv4 no ipv6. Maybe this could be a problem? I thought about disabling the open Matrix port for the ipv6?

**Comment by KeenMaron** [comment]
Update: 
I typed the matrix server adress via ipv4 -> direct response:
![image](https://user-images.githubusercontent.com/52194699/144645015-4027fe48-c44c-4de9-a415-cc5ff72fa486.png)

via ipv6 the site is loading ~20 seconds -> connection error

maybe it is a ipv6 / ipv4 / ipv6 to ipv4 fallback issue?

**Comment by KeenMaron** [comment]
Update: I disabled ipv6 at my raspberry pi. All my services are still working over ipv4, except matrix. If i try to connect via ipv4 like before i get a bad gateway. Sounds like a nginx missconfiguration? After enabling ipv6 again i can reach matrix over the ipv4 adress.

**Comment by ouchadam** [comment]
sounds very strange :thinking: networking isn't my area of expertise I'm afraid, the #synapse-admins:matrix.org room might be a better source of information

out of interest does overriding the mobile network dns have any effect? https://www.androidpolice.com/2020/03/26/make-android-use-dns-server-choice/

System settings -> connections -> More connection settings -> Private DNS -> dns.google

**Comment by KeenMaron** [comment]
I tried dns.google and 1.1.1.1 dns. No change.

**verbose logs** -> I will check it now

**Comment by KeenMaron** [comment]
@ouchadam I just have sent the bug report. I activated the verbose log, logged in via wifi, closed the app, logged in via mobile network. Can you check if you can see the report?

**Comment by KeenMaron** [comment]
Update: 
I finally fixed my ipv6. I can access now matrix via my dyndns / directly via ipv4 / ipv6. After that, there is no delay anymore if i connect with mobile network. It connects instantly (no 2-3 seconds delay).
My question is, was this a problem of my configuration? Or should element android client handle this kind of situation if ipv6 is not reachable, that it should switch faster to ipv4? FluffyChat worked without delay if ipv6 wasn't reachable (although my dyndns suggested it works).

**Comment by ouchadam** [comment]
great to hear you've got it working and thanks for all the updates :100: 

I can confirm the logs came through and they also show the sync request taking up to 60 seconds, other times it was near instant.

The difference with `FluffyChat` may be the networking library we use which will attempt to use multiple IPs if possible when a connection error is detected. I suspect the library was attempting to connect via IPv6 multiple times, failing and then trying IPv4 and succeeding.

> OkHttp perseveres when the network is troublesome: it will silently recover from common connection problems. If your service has multiple IP addresses, OkHttp will attempt alternate addresses if the first connect fails. This is necessary for IPv4+IPv6 and services hosted in redundant data centers

https://square.github.io/okhttp/

This may help improve the IPv6 to IPv4 fallback timings in the future https://github.com/square/okhttp/issues/506

In this case I'd lean towards closing this issue and push the responsibility of IP routing to the homeserver admins (I don't think it's wrong that the android client defaults to IPv6 if the DNS tells us the address exists) 

----

To sum up - _it looks like_~ this issue was caused by a listed but misconfigured IPv6 entry which the android app attempted to connect to, eventually falling back to the IPv4 address (taking up 60 seconds in the process~)

**Comment by KeenMaron** [comment]
The explenation makes sense, thank you :) !
